### PR TITLE
dolomite fixes

### PR DESCRIPTION
**Pull request by cdoern**
when additional_args defaults change, overwrite file

    currently new args like use_dolomite don't propogate into the additional args file since on most systems
    it already exists. Check if the file contents match the defaults. If they do not, overwrite

    make use_dolomite mandatory

    if a user deletes use_dolomite from cfg, we shouldn't run training successfully

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/configuration.py [hunk 1]
```diff
@@ -1378,8 +1383,10 @@ def finish_additional_train_args(current_additional):
         # user has not run `ilab config init`, yet. This should only happen once
     _expand_paths(additional_args_and_defaults)
     for key, val in additional_args_and_defaults.items():
+        print(key, val)
         if key not in current_additional:
             current_additional[key] = val
+    print(current_additional)
```
Comment (jaideepr97) [inline src/instructlab/configuration.py:1389]: ditto

File: src/instructlab/cli/model/train.py [hunk 2]
```diff
@@ -245,6 +245,7 @@ def clickpath_setup(is_dir: bool) -> click.Path:
     cls=clickext.ConfigOption,
     config_sections=ADDITIONAL_ARGUMENTS,
     type=click.BOOL,
+    required=True,
```
Comment (jaideepr97) [inline src/instructlab/cli/model/train.py:248]: could you add the comment saying ` # default from config` to keep it consistent with other additonal_args?

File: src/instructlab/configuration.py [hunk 3]
```diff
@@ -1142,8 +1142,13 @@ def ensure_storage_directories_exist() -> bool:
...
 
-    # create expert_args file for users to see/edit
-    if not os.path.isfile(DEFAULTS.TRAIN_ADDITIONAL_OPTIONS_FILE):
+    # open file in read mode to load current contents
+    with open(DEFAULTS.TRAIN_ADDITIONAL_OPTIONS_FILE, "r", encoding="utf-8") as infile:
+        current_dict = yaml.load(infile) or {}
+
+    # check if if there are new keys in defaults
+    if not DEFAULTS.ADDITIONAL_ARGS_DEFAULTS.keys() <= current_dict.keys():
+        # open file in write mode and update contents if there are new keys
```
Comment (jaideepr97) [inline src/instructlab/configuration.py:1151]: why not just always overwrite this file with the defaults?

Comment (cdoern) [reply src/instructlab/configuration.py:1151]: I'd like to minimize writes if the file exists. This gets run alot

File: src/instructlab/configuration.py (hunk 1 above)
Comment (cdoern) [reply src/instructlab/configuration.py]: Done

File: src/instructlab/configuration.py [hunk 4]
```diff
@@ -1378,8 +1383,10 @@ def finish_additional_train_args(current_additional):
         # user has not run `ilab config init`, yet. This should only happen once
     _expand_paths(additional_args_and_defaults)
     for key, val in additional_args_and_defaults.items():
+        print(key, val)
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/configuration.py]: Done

File: src/instructlab/configuration.py (hunk 3 above)
Comment (jaideepr97) [reply src/instructlab/configuration.py:1151]: should the file read be in a try/except or will it properly set `current_dict` to {} if the file doesn't exist?

Comment (cdoern) [reply src/instructlab/configuration.py:1151]: current_dict will be {} if it doesn't exist I believe

File: src/instructlab/configuration.py [hunk 5]
```diff
@@ -1142,8 +1145,20 @@ def ensure_storage_directories_exist() -> bool:
...
+    current_dict: dict[str, Any] = {}
+    try:
+        with open(
+            DEFAULTS.TRAIN_ADDITIONAL_OPTIONS_FILE, "r", encoding="utf-8"
+        ) as infile:
+            current_dict = yaml.load(infile) or {}
+    except FileNotFoundError:
+        logger.debug("Additional Arguments internal file not found, creating now")
+
+    # check if if there are new keys in defaults
```
Comment (danmcp) [inline src/instructlab/configuration.py:1159]: nit: if if

**Review by nathan-weinberg — APPROVED** [review]

File: src/instructlab/configuration.py [hunk 6]
```diff
@@ -1142,8 +1145,20 @@ def ensure_storage_directories_exist() -> bool:
...
+    try:
+        with open(
+            DEFAULTS.TRAIN_ADDITIONAL_OPTIONS_FILE, "r", encoding="utf-8"
+        ) as infile:
+            current_dict = yaml.load(infile) or {}
+    except FileNotFoundError:
+        logger.debug("Additional Arguments internal file not found, creating now")
+
+    # check if there are new keys in defaults
+    if DEFAULTS.ADDITIONAL_ARGS_DEFAULTS.keys() > current_dict.keys():
```
Comment (khaledsulayman) [inline src/instructlab/configuration.py:1160]: nit: prefer using `len()` here to be explicit

File: src/instructlab/configuration.py [hunk 7]
```diff
@@ -1142,8 +1145,20 @@ def ensure_storage_directories_exist() -> bool:
 
     fresh_install = recreate_system_profiles()
 
-    # create expert_args file for users to see/edit
-    if not os.path.isfile(DEFAULTS.TRAIN_ADDITIONAL_OPTIONS_FILE):
+    # open file in read mode to load current contents
+
+    current_dict: dict[str, Any] = {}
```
Comment (khaledsulayman) [inline src/instructlab/configuration.py:1150]:
```suggestion
    train_options_dict: dict[str, Any] = {}
```

**Review by khaledsulayman — COMMENTED** [review]
looks good. 2 nits but won't block

**Review by alinaryan — APPROVED** [review]

**Pull request merged**